src/Servers/Kestrel/Core/src/CoreStrings.resx

```diff
@@ -590,4 +590,10 @@ For more information on configuring HTTPS see https://go.microsoft.com/fwlink/?l
   <data name="Http2ErrorStreamAborted" xml:space="preserve">
     <value>A frame of type {frameType} was received after stream {streamId} was reset or aborted.</value>
   </data>
+  <data name="OnCompletedNotSupported" xml:space="preserve">
+    <value>OnCompleted cannot be registered during a call to OnCompleted.</value>
+  </data>
+  <data name="OnStartingNotSupported" xml:space="preserve">
+    <value>OnStarting cannot be registered during a call to OnStarting.</value>
```

Comment:
If you decide to go back to throwing, we could do stricter checks with the following messages. "OnCompleted cannot be registered after the response completed" and "OnStarting cannot be registered after the response has started".

It would be nice for OnCompleted to guarantee that the callback will eventually get called if it doesn't throw. Even after this change, it's possible for OnCompleted callbacks to get cleared without being called.

Reply:
Not in this PR. File an issue.